src/inmanta/data/__init__.py

```diff
@@ -2191,24 +2191,28 @@ def to_dict(self) -> JsonType:
 
     @classmethod
     async def execute_in_retryable_transaction(
-        cls, fnc: Callable[[Connection], Awaitable[TransactionResult]], tx_isolation_level: Optional[str] = None
+        cls,
+        fnc: Callable[[Connection], Awaitable[TransactionResult]],
+        tx_isolation_level: Optional[str] = None,
     ) -> TransactionResult:
         """
         Execute the queries in fnc using the transaction isolation level `tx_isolation_level` and return the
         result returned by fnc. This method performs retries when the transaction is aborted due to a
         serialization error.
         """
         async with cls.get_connection() as postgresql_client:
-            max_retries = 3
+            attempt = 1
             while True:
                 try:
                     async with postgresql_client.transaction(isolation=tx_isolation_level):
                         return await fnc(postgresql_client)
                 except SerializationError:
-                    if max_retries >= 0:
-                        max_retries -= 1
+                    if attempt > 3:
+                        raise Exception("Failed to execute transaction after 3 attempts.")
                     else:
-                        raise Exception("Failed to execute transaction after 3 retries.")
+                        # Exponential backoff
+                        await asyncio.sleep(10 * pow(2, attempt))
```

Comment:
20-80 seconds seems like a lot to me. I was expecting more something like 0.1-2 seconds or so. But I must admit I can't say with confidence what would be an appropriate interval here.

Reply:
I put it to 10ms, 100ms, 1000ms

Reply:
Looks good to me.